modules/compactor/compactor.go

```diff
@@ -45,50 +56,94 @@ func New(cfg Config, store storage.Store, overrides *overrides.Overrides, reg pr
 		overrides: overrides,
 	}
 
-	subservices := []services.Service(nil)
 	if c.isSharded() {
-		lifecyclerCfg := c.cfg.ShardingRing.ToLifecyclerConfig()
-		lifecycler, err := ring.NewLifecycler(lifecyclerCfg, ring.NewNoopFlushTransferer(), "compactor", cfg.OverrideRingKey, false, log.Logger, prometheus.WrapRegistererWithPrefix("cortex_", reg))
+		reg = prometheus.WrapRegistererWithPrefix("cortex_", reg)
+
+		lifecyclerStore, err := kv.NewClient(
+			cfg.ShardingRing.KVStore,
+			ring.GetCodec(),
+			kv.RegistererWithKVName(reg, ring.CompactorRingKey+"-lifecycler"),
+			log.Logger,
+		)
 		if err != nil {
-			return nil, errors.Wrap(err, "unable to initialize compactor ring lifecycler")
+			return nil, err
 		}
-		c.ringLifecycler = lifecycler
-		subservices = append(subservices, c.ringLifecycler)
 
-		ring, err := ring.New(lifecyclerCfg.RingConfig, "compactor", cfg.OverrideRingKey, log.Logger, prometheus.WrapRegistererWithPrefix("cortex_", reg))
+		delegate := ring.BasicLifecyclerDelegate(c)
+		delegate = ring.NewLeaveOnStoppingDelegate(delegate, log.Logger)
+		delegate = ring.NewAutoForgetDelegate(ringAutoForgetUnhealthyPeriods*cfg.ShardingRing.HeartbeatTimeout, delegate, log.Logger)
+
+		bcfg, err := toBasicLifecyclerConfig(cfg.ShardingRing, log.Logger)
 		if err != nil {
-			return nil, errors.Wrap(err, "unable to initialize compactor ring")
+			return nil, err
 		}
-		c.Ring = ring
-		subservices = append(subservices, c.Ring)
 
-		c.subservices, err = services.NewManager(subservices...)
+		c.ringLifecycler, err = ring.NewBasicLifecycler(bcfg, ring.CompactorRingKey, cfg.OverrideRingKey, lifecyclerStore, delegate, log.Logger, reg)
 		if err != nil {
-			return nil, fmt.Errorf("failed to create subservices %w", err)
+			return nil, errors.Wrap(err, "unable to initialize compactor ring lifecycler")
+		}
+
+		c.Ring, err = ring.New(c.cfg.ShardingRing.ToLifecyclerConfig().RingConfig, ring.CompactorRingKey, cfg.OverrideRingKey, log.Logger, reg)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to initialize compactor ring")
 		}
-		c.subservicesWatcher = services.NewFailureWatcher()
-		c.subservicesWatcher.WatchManager(c.subservices)
 	}
 
 	c.Service = services.NewBasicService(c.starting, c.running, c.stopping)
 
 	return c, nil
 }
 
-func (c *Compactor) starting(ctx context.Context) error {
-	if c.subservices != nil {
+func (c *Compactor) starting(ctx context.Context) (err error) {
+	// In case this function will return error we want to unregister the instance
+	// from the ring. We do it ensuring dependencies are gracefully stopped if they
+	// were already started.
+	defer func() {
+		if err == nil || c.subservices == nil {
```

Comment:
i had never though of checking err in a defer. this is cool

Reply:
Shamelessly taken from cortex :)